File: backend/utils/validators.py
"""Input validation helpers."""
import re
from utils.constants import OBJECTIVE_MAP, CTA_OPTIONS


def validate_ad_account_id(account_id: str) -> str:
    """Ensure ad account ID is in the correct format."""
    cleaned = account_id.strip()
    if not cleaned.startswith("act_"):
        cleaned = f"act_{cleaned}"
    if not re.match(r"^act_\d+$", cleaned):
        raise ValueError(f"Invalid ad account ID format: {account_id}")
    return cleaned


def validate_objective(objective: str) -> str:
    """Map and validate campaign objective."""
    lower = objective.lower().strip()
    if lower in OBJECTIVE_MAP:
        return OBJECTIVE_MAP[lower]
    if objective.startswith("OUTCOME_"):
        return objective
    raise ValueError(
        f"Unknown objective: {objective}. "
        f"Valid: {list(OBJECTIVE_MAP.keys())}"
    )


def validate_cta(cta: str) -> str:
    """Validate CTA button type."""
    upper = cta.upper().strip()
    if upper in CTA_OPTIONS:
        return upper
    raise ValueError(f"Unknown CTA: {cta}. Valid: {CTA_OPTIONS}")


def extract_drive_folder_id(url: str) -> str:
    """Extract Google Drive folder ID from URL."""
    patterns = [
        r"folders/([a-zA-Z0-9_-]+)",
        r"id=([a-zA-Z0-9_-]+)",
    ]
    for pattern in patterns:
        match = re.search(pattern, url)
        if match:
            return match.group(1)
    raise ValueError(f"Could not extract folder ID from URL: {url}")

File: api/lib/utils/validators.ts
import { OBJECTIVE_MAP, CTA_OPTIONS } from "./constants.js";

/**
 * Ensure ad account ID is in the correct format (act_XXXXXXXXX).
 * Prepends "act_" if missing, then validates the format.
 */
export function validateAdAccountId(accountId: string): string {
  let cleaned = accountId.trim();
  if (!cleaned.startsWith("act_")) {
    cleaned = `act_${cleaned}`;
  }
  if (!/^act_\d+$/.test(cleaned)) {
    throw new Error(`Invalid ad account ID format: ${accountId}`);
  }
  return cleaned;
}

/**
 * Map and validate a campaign objective string to its ODAX API value.
 * Accepts both friendly names ("sales", "traffic") and raw API values ("OUTCOME_SALES").
 */
export function validateObjective(objective: string): string {
  const lower = objective.toLowerCase().trim();
  if (lower in OBJECTIVE_MAP) {
    return OBJECTIVE_MAP[lower];
  }
  if (objective.startsWith("OUTCOME_")) {
    return objective;
  }
  const validKeys = Object.keys(OBJECTIVE_MAP).join(", ");
  throw new Error(
    `Unknown objective: ${objective}. Valid options: ${validKeys}`
  );
}

/**
 * Validate a CTA (call-to-action) button type against the allowed list.
 */
export function validateCta(cta: string): string {
  const upper = cta.toUpperCase().trim();
  if (CTA_OPTIONS.includes(upper)) {
    return upper;
  }
  throw new Error(
    `Unknown CTA: ${cta}. Valid options: ${CTA_OPTIONS.join(", ")}`
  );
}

/**
 * Extract a Google Drive folder ID from a URL.
 * Supports standard folder URLs and id= query parameter format.
 */
export function extractDriveFolderId(url: string): string {
  const patterns = [/folders\/([a-zA-Z0-9_-]+)/, /id=([a-zA-Z0-9_-]+)/];
  for (const pattern of patterns) {
    const match = url.match(pattern);
    if (match) {
      return match[1];
    }
  }
  throw new Error(`Could not extract folder ID from URL: ${url}`);
}
